### PR TITLE
[None][perf] Cache FMHA library selection in TrtllmAttention

### PR DESCRIPTION
**Pull request by yuxianq**
## Description

Cache FMHA library selection per `TrtllmAttention` instance so eager forwards
do not repeatedly evaluate every library's `is_supported` predicate. The cache
is scoped to one attention layer and is reset whenever its ordered FMHA library
list is rebuilt.

The cache key contains the request-level inputs that can change selection:

- normalized source context and generation batch sizes
- normalized generation Q sequence length
- effective `AttentionMaskType`
- speculative-decoding state
- output and output-scale dtypes

Explicit custom masks and any non-`None` `attention_mask_data` are represented
as `AttentionMaskType.custom_mask`. Output dtypes keep LoRA's packed and
unpacked output representations in separate entries.

Batch sizes and generation Q length use ceiling-normalized grids based on
TRTLLM-Gen JIT warmup candidates. Each candidate's predecessor is also a grid
point so a support boundary at that candidate cannot alias with the interval
below it; values above the largest point share the largest bucket. For size
bucketing, context selection uses only batch size, while generation also uses
Q length. The source context/generation batch counts are retained for mixed
inputs and compacted MLA phase calls, with a zero generation Q length
distinguishing a context-only subcall. Generation Q lengths are uniform in the
executor, including padded speculative-decoding batches.

Only successful selections are cached. Each key retains an independent FMHA
selection that later misses cannot mutate. Rebuilding the FMHA library list
clears both the selection cache and diagnostic snapshots. Cache lookup is
bypassed globally while autotuning is active so tuning-time policy changes
cannot enter the serving cache. The `Fmha.is_supported` contract now requires
new forward-varying selection inputs to be represented in the key and new size
boundaries to be represented in the corresponding grid.

For validation, setting `TRTLLM_FMHA_CACHE_SANITY_CHECK=1` before engine
initialization or an FMHA library rebuild re-runs uncached selection on every
cache hit. A mismatch logs the cached and uncached selections, diffs captured
forward inputs, and raises an error. Snapshots retain tensor shapes plus
primitive, enum, dtype, and device values without retaining tensors or the KV
cache manager; private and derived properties are excluded.

This reduces Python forward overhead in eager execution and allows more
detailed support checks without paying their full cost on every layer and
forward.

## Performance Results

Profiled on an NVIDIA B200 (SM100) with TinyLlama-1.1B BF16, 16 input tokens,
128 decode steps, beam width 2, and KV-cache page size 64. CUDA graphs, overlap
scheduling, torch compilation, chunked prefill, and KV block reuse were
disabled. Each of the 22 attention layers had the ordered backend list
`FlashInferTrtllmGenFmha -> FallbackFmha`; FlashInfer rejected beam-search
generation and every layer selected fallback.

A same-engine A/B compared cached selection with a direct uncached
`_select_non_mla_fmha` call. Timing used 20 randomized paired blocks with two
requests per block.

| Metric | Uncached | Cached | Change |
|---|---:|---:|---:|
| Mean request latency | 1426.86 ms | 1389.50 ms | -37.37 ms (-2.62%) |
| Median request latency | 1427.83 ms | 1387.96 ms | -39.86 ms |
| Mean latency per decode step | 11.147 ms | 10.855 ms | -0.292 ms |
| Throughput speedup | - | - | 1.0269x |

The paired bootstrap 95% confidence interval was 34.66-40.05 ms for latency
reduction and 1.0249x-1.0288x for throughput speedup.

In a separate 64-step attribution run, uncached selection made 1,408
FlashInfer and 1,408 fallback `is_supported` calls. A cold cache made 44 calls
to each (the first context and generation selections for 22 layers), while a
steady cache made zero. Direct `is_supported` time was 18.20 ms, or about
284 us per decode step, closely matching the uninstrumented 292 us per-step
saving.

The uncached arm directly called `_select_non_mla_fmha`, omitting the old
path's inexpensive MLA branch, so the measured improvement is conservative.

## Test Coverage

- `pre-commit run --files tensorrt_llm/_torch/attention_backend/trtllm.py tensorrt_llm/_torch/attention_backend/fmha/interface.py tests/unittest/_torch/attention/test_combined_fmha.py`
- `python3 -m py_compile tensorrt_llm/_torch/attention_backend/trtllm.py tensorrt_llm/_torch/attention_backend/fmha/interface.py tests/unittest/_torch/attention/test_combined_fmha.py`
- `git diff --check origin/main...HEAD`
- Added behavioral unit coverage for grid-boundary separation, same-cell reuse,
  context batch-only size dispatch, compacted MLA phase isolation, custom-mask
  data, speculative decoding, LoRA output representation, independent combined
  selections, failed selections, cache invalidation, and global autotuning
  bypass.
- Added sanity-check coverage for equivalent combined selections and mismatch
  diagnostics with forward-input differences.
- Focused pytest execution was unavailable in the local worktree because the
  compiled `tensorrt_llm.bindings` module is not installed.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/fmha/interface.py [hunk 1]
```diff
@@ -109,7 +109,12 @@ def is_supported(
         *,
         phase: Optional[FmhaPhase] = None,
     ) -> bool:
-        """Return whether this library supports the request or requested phase."""
+        """Return whether this library supports the request or requested phase.
+
+        Forward-varying selection conditions must be represented in
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/fmha/interface.py:114]: If so, is it possible that we should just pass in `_FmhaSelectionCacheKey` as input, and by doing so only allowing it to see the value changes in the cache key?

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 2]
```diff
@@ -51,6 +52,85 @@
...
+    256,
+    384,
+    512,
+    768,
+    1024,
+    1280,
+    1536,
+    2048,
+)
+_FMHA_SELECTION_SEQ_LEN_Q_GRID: Tuple[int, ...] = (1, 2, 4, 8, 32, 64, 128)
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: The batch/seqlen_q grid may not match what's in cutedsl backend

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 3]
```diff
@@ -51,6 +52,85 @@
...
+
+
+class _FmhaSelectionCacheKey(NamedTuple):
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: Maybe missing some values that can change dynamically, like mask type (may change for multimodal models depending on whether request contains image token etc.)

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 2 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: It has covered how we select between cutedsl mla fmha and fallback fmha, do you mean it mismatches with heuristics inside cutedsl mla fmha? We still use the existed heuristics to route inside fmha lib, so it is unrelated.

File: tensorrt_llm/_torch/attention_backend/fmha/interface.py (hunk 1 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/fmha/interface.py:114]: We also need other values to implement is_supported. E.g., `CuteDslMlaFmha` checks `metadata.beam_width` in `is_supported` since it only supports `beam_width == 1`, but `beam_width` is unchanged for a LLM instance, so we don't need to add it to cache key.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 3 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: Scan all fields and add attention_mask_type/use_spec_decoding into cache key, which could change per request and lead to different fmha selection.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 2 above)
Comment (pengbowang-nv) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: For cutedsl mla, see https://github.com/yuxianq/TensorRT-LLM/blob/edf35d334cab9047f7684433f53a18b628a5945e/tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py#L272 where we use min_batch_size to filter and num_heads/seq_len_q as hash key:

1. if numheads=16, for seqlen q = 3, we will use (128, 4) result while without cache (128, 3) will be rejected.
2. if numheads=128 and seqlen = 1, if batchsize in 56-64, with/without cache the backend selction will be different.

Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: Fixed. Now a grid [a, b] will be split into three range, a, [a+1, b-1], b, so that the shape between a and b will not automatically reuse the cache key of a or b, so that 56-64 can have separate fmha libs.

Comment (pengbowang-nv) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: This indeed resolved the above problem. 

Yet here we are making an assumption that all backend changing parameter turing point should fall on our grid, for the cache to be correct. Then a small NIT is: we don't need to have a dense grid as in TRTLLM-Gen attention. TRTLLM-Gen attention used that dense grid because it's hard to know where turning point is, and won't cause correctness issue. But the logic to select backend is simpler, and we are obliged to keep the grid consistent with the selection logic (otherwise maybe incorrect). So we can use a grid that match our backend selection logic, and avoid giving a long list and pretending we can handle general case.

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 4]
```diff
@@ -44,13 +48,252 @@
...
+    # Keep all temporary selections out of the serving cache so future FMHAs
+    # inherit the same cache boundary.
+    from tensorrt_llm._torch.autotuner import AutoTuner
+
+    autotuner = AutoTuner._instance
+    return autotuner is None or not autotuner.is_tuning_mode
+
+
+def _is_fmha_cache_sanity_check_enabled() -> bool:
+    return os.environ.get(_FMHA_CACHE_SANITY_CHECK_ENV, "0") == "1"
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/trtllm.py:170]: Should follow trtllm env varible convention and use `TRTLLM_xxx`

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 5]
```diff
@@ -1765,8 +2017,123 @@ def create_fmha_libs(self) -> None:
...
+        subcall from a generation-active subcall.
+        """
+        attention_input_type = forward_args.attention_input_type
+        output_dtype = (forward_args.output.dtype
+                        if forward_args.output is not None else None)
+        output_sf_dtype = (forward_args.output_sf.dtype
+                           if forward_args.output_sf is not None else None)
+        # Explicit mask data has the same FMHA support constraints as a
+        # custom mask enum, even when the accompanying enum remains causal.
+        attention_mask_type = (AttentionMaskType.custom_mask if (
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: The target branch has added TritonCustomMaskFmha and new FallbackFmha support predicates since this PR’s merge base. The new selectors test those conditions separately. Therefore `CAUSAL + mask_data`, `CUSTOM + mask_data`, and `CUSTOM + None` can share a cache key but produce different uncached selections.

File: tests/unittest/_torch/attention/test_combined_fmha.py [hunk 6]
```diff
@@ -379,3 +399,777 @@ def test_flashinfer_context_fallback_scope(
...
+) -> SimpleNamespace:
+    return SimpleNamespace(
+        num_contexts=num_contexts,
+        num_generations=num_generations,
+        num_ctx_tokens=num_ctx_tokens,
+        use_spec_decoding=use_spec_decoding,
+    )
+
+
+def test_fmha_cache_sanity_check_env(
```
Comment (pengbowang-nv) [inline tests/unittest/_torch/attention/test_combined_fmha.py]: I suggest remove below tests
- `test_fmha_cache_sanity_check_env`
- `test_fmha_cache_input_snapshot_excludes_manager_and_properties`
- `test_fmha_cache_key_tracks_use_spec_decoding`
- `test_fmha_cache_key_tracks_attention_mask_type`
- `test_fmha_cache_key_distinguishes_compacted_phases`

and there might be more cases that are simple existing code written in forms of assert, I just listed here some obvious ones.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 2 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: If we use a grid to exactly match the current selection logic, we take risk to wrongly cache fmha when a new fmha lib is added and slightly adjust the selection logic. So I keep the grid as general as possible here.

File: tests/unittest/_torch/attention/test_combined_fmha.py (hunk 6 above)
Comment (yuxianq) [reply tests/unittest/_torch/attention/test_combined_fmha.py]: Addressed in c4d27226d4. I removed all five listed direct helper/key tests plus four additional implementation-detail tests. The speculative-decoding and compacted-MLA cases are now behavioral selection tests that run both request orders and verify cache hits do not reuse the wrong FMHA; they no longer assert key fields directly. Pre-commit, Python compilation, and diff checks pass.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 4 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/trtllm.py:170]: The variable already follows the requested `TRTLLM_*` convention: its exact name is `TRTLLM_FMHA_CACHE_SANITY_CHECK`. Did you have a different naming change in mind?

Comment (pengbowang-nv) [reply tensorrt_llm/_torch/attention_backend/trtllm.py:170]: Sorry I missed it. It's a false positive. 🫠

**Review by pengbowang-nv — APPROVED** [review]
LGTM, thanks!

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 4 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/trtllm.py:170]: The env variable has used `TRTLLM` prefix, `_FMHA_CACHE_SANITY_CHECK_ENV ` is just a private constant name.

File: tests/unittest/_torch/attention/test_combined_fmha.py (hunk 6 above)
Comment (yuxianq) [reply tests/unittest/_torch/attention/test_combined_fmha.py]: Removed.
